### PR TITLE
AGENT-357 Port luadata to octeon-linux

### PR DESCRIPTION
**Pull request by marcelstanley**

### CODE REVIEW

File: binary.c [hunk 1]
```diff
@@ -132,13 +136,20 @@ swap_bytes_out(uint64_t *value, size_t width)
...
+	byte_t mask;
+
+	msb_offset = MSB_OFFSET(offset);
+	lsb_offset = LSB_OFFSET(width, msb_offset);
+	pos        = BYTE_POSITION(offset);
+	overflow   = OVERFLOW_BITS(width, msb_offset, lsb_offset);
+	mask       = MASK(msb_offset, lsb_offset);
 
-	uint64_t value = (byte_t) (bytes[ pos ] & mask) >> lsb_offset;
+	value      = (byte_t) (bytes[ pos ] & mask) >> lsb_offset;
```
Comment (tammela) [inline binary.c]:
```C
(byte_t) (bytes[pos] & mask) >> lsb_offset;
```

File: sys/endian.h [hunk 2]
```diff
@@ -40,4 +40,18 @@
 #define bswap64 __builtin_bswap64
 #endif
 
-#endif /* _ENDIAN_H_ */
\ No newline at end of file
+#ifdef __MIPSEL__
+#define __BYTE_ORDER __LITTLE_ENDIAN
```
Comment (tammela) [inline sys/endian.h]:
```C
#define __BYTE_ORDER (__LITTLE_ENDIAN)
```
And so on...

File: luadata.c [hunk 3]
```diff
@@ -327,27 +330,22 @@ ldata_topointer(lua_State *L, int index, size_t *size)
...
 
-MODULE(MODULE_CLASS_MISC, luadata, "lua");
+#ifdef _KERNEL
+EXPORT_SYMBOL(luaopen_data);
 
-static int
-luadata_modcmd(modcmd_t cmd, void *opaque)
+static int __init
+data_init(void)
+__attribute__ ((unused));
```
Comment (tammela) [inline luadata.c]: I don't think this is needed, more info?

File: luadata.h [hunk 4]
```diff
@@ -29,8 +29,7 @@
 #define _LUA_DATA_H_
```
Comment (tammela) [inline luadata.h]: Add CUJO copyright

File: sys/endian.h (hunk 2 above)
Comment (lneto) [inline sys/endian.h]: I think byte order should be defined by the kernel.. we just need to figure out the define..

File: sys/endian.h [hunk 5]
```diff
@@ -40,4 +40,18 @@
...
+#else
+#define __BYTE_ORDER __BIG_ENDIAN
+#endif
+
+#ifdef _KERNEL
+#define __LITTLE_ENDIAN 1234
+#define __BIG_ENDIAN 4321
```
Comment (lneto) [inline sys/endian.h]: this should be defined somewhere also..

File: sys/endian.h [hunk 6]
```diff
@@ -40,4 +40,18 @@
...
+#define BIG_ENDIAN      __BIG_ENDIAN
+#define LITTLE_ENDIAN   __LITTLE_ENDIAN
+#define BYTE_ORDER      __BYTE_ORDER
```
Comment (lneto) [inline sys/endian.h]: don't need this indirection.. if needed, we can use the makefile for it..

File: data.c [hunk 7]
```diff
@@ -178,24 +188,15 @@ data_new(lua_State *L, void *ptr, size_t size, bool free)
...
 int
 data_new_segment(lua_State *L, data_t *data, size_t offset, size_t length)
 {
+        handle_t *handle; 
+        handle  = data->handle;
+
 	if (!check_limits(data, offset, length))
 		return 0;
 
-	handle_t *handle = data->handle;
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 8]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
 		return;
 
-	size_t len;
-	const char *s = lua_tolstring(L, value_ix, &len);
+	s = lua_tolstring(L, value_ix, &len);
 	if (s == NULL)
 		return;
 
-	char *ptr = (char *) data_get_ptr(data); 
+	ptr = (char *) data_get_ptr(data);
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 9]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
+	size_t len;
+        const char *s; 
+        char *ptr; 
+
 	if (!check_str_limits(data, entry))
 		return;
 
-	size_t len;
-	const char *s = lua_tolstring(L, value_ix, &len);
+	s = lua_tolstring(L, value_ix, &len);
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 10]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
 		return;
 
-	byte_t *ptr = (byte_t *) data_get_ptr(data);
+	ptr = (byte_t *) data_get_ptr(data);
 	if (ptr == NULL)
 		return;
 
 	/* assertion: LUA_INTEGER_BIT <= 64 */
-	lua_Integer value = lua_tointeger(L, value_ix);
+	value = lua_tointeger(L, value_ix);
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 11]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
 set_num(lua_State *L, data_t *data, layout_entry_t *entry, int value_ix)
 {
+        byte_t *ptr;
+        lua_Integer value;
+
 	if (!check_num_limits(data, entry))
 		return;
 
-	byte_t *ptr = (byte_t *) data_get_ptr(data);
+	ptr = (byte_t *) data_get_ptr(data);
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 12]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
+        const char *ptr;
+        const char *s;
+
 	if (!check_str_limits(data, entry))
 		return 0;
 
-	const char *ptr = (const char *) data_get_ptr(data); 
-	const char *s = ptr + entry->offset;
+	ptr = (const char *) data_get_ptr(data); 
+	s = ptr + entry->offset;
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 13]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
 		return 0;
 
-	byte_t *ptr = (byte_t *) data_get_ptr(data);
+	ptr = (byte_t *) data_get_ptr(data);
 	if (ptr == NULL)
 		return 0;
 
 	/* assertion: LUA_INTEGER_BIT <= 64 */
-	lua_Integer value = binary_get_uint64(BINARY_PARMS(data, entry, ptr));
+	value = binary_get_uint64(BINARY_PARMS(data, entry, ptr));
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 14]
```diff
@@ -115,58 +113,70 @@ new_data(lua_State *L, handle_t *handle, size_t offset, size_t length)
...
 get_num(lua_State *L, data_t *data, layout_entry_t *entry)
 {
-	if (!check_num_limits(data, entry))
+	byte_t *ptr;
+        lua_Integer value;
+
+        if (!check_num_limits(data, entry))
 		return 0;
 
-	byte_t *ptr = (byte_t *) data_get_ptr(data);
```
Comment (lneto) [inline data.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: data.c [hunk 15]
```diff
@@ -28,11 +28,9 @@
 #ifndef _KERNEL
 #include <string.h>
 #else
-#include <lib/libkern/libkern.h>
+#define MIN	min
```
Comment (lneto) [inline data.c]: we can use the makefile for that..

File: binary.h [hunk 16]
```diff
@@ -28,16 +28,15 @@
...
 #ifdef _KERNEL
-#include <sys/types.h>
+#include <linux/types.h>
 #else
 #include <stddef.h>
 #include <stdint.h>
 #endif
 
-#define BYTE_BIT	CHAR_BIT
+#define CHAR_BIT	(8)
```
Comment (lneto) [inline binary.h]: I think it should be defined somewhere.. if not, use the makefile

File: binary.c [hunk 17]
```diff
@@ -161,11 +172,17 @@ void
...
+        size_t lsb_offset;
+        size_t pos;
+        size_t overflow;
+        byte_t clear_mask;
+
+	msb_offset = MSB_OFFSET(offset);
+	lsb_offset = LSB_OFFSET(width, msb_offset);
+	pos        = BYTE_POSITION(offset);
+	overflow   = OVERFLOW_BITS(width, msb_offset, lsb_offset);
+	clear_mask = ~MASK(msb_offset, lsb_offset);
```
Comment (lneto) [inline binary.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: binary.c (hunk 1 above)
Comment (lneto) [inline binary.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: binary.c [hunk 18]
```diff
@@ -132,13 +136,20 @@ swap_bytes_out(uint64_t *value, size_t width)
...
-	size_t lsb_offset = LSB_OFFSET(width, msb_offset);
-	size_t pos        = BYTE_POSITION(offset);
-	size_t overflow   = OVERFLOW_BITS(width, msb_offset, lsb_offset);
-	byte_t mask       = MASK(msb_offset, lsb_offset);
+	size_t msb_offset;
+	size_t lsb_offset;
+	size_t pos;
+	size_t overflow;
+	int64_t value;
+	byte_t mask;
```
Comment (lneto) [inline binary.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: binary.c [hunk 19]
```diff
@@ -114,7 +118,7 @@ swap_bytes_out(uint64_t *value, size_t width)
 #define MSB_OFFSET(offset)		(offset % BYTE_BIT)
 
 #define LSB_OFFSET(width, msb_offset)					\
-	MAX((ssize_t) (BYTE_BIT - msb_offset - width), 0)
+	MAX((ssize_t) (BYTE_BIT - msb_offset - width), (ssize_t) 0)
```
Comment (lneto) [inline binary.c]: do we have a warning on this? Oo

File: binary.c [hunk 20]
```diff
@@ -86,25 +88,27 @@ contract(uint64_t *value, size_t width, size_t msb_offset, byte_t truncated)
...
 static void
 swap_bytes_out(uint64_t *value, size_t width)
 {
+	byte_t truncated;
 	size_t msb_offset = VALUE_MSB_OFFSET(width);
 
 	*value = bswap64(*value);
 
-	byte_t truncated = TRUNCATED_BITS(width);
+	truncated = TRUNCATED_BITS(width);
```
Comment (lneto) [inline binary.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: binary.c [hunk 21]
```diff
@@ -86,25 +88,27 @@ contract(uint64_t *value, size_t width, size_t msb_offset, byte_t truncated)
...
 {
+	byte_t truncated;
 	size_t msb_offset = VALUE_MSB_OFFSET(width);
 
 	*value <<= msb_offset;
 
 	*value = bswap64(*value);
 
-	byte_t truncated = TRUNCATED_BITS(width);
+	truncated = TRUNCATED_BITS(width);
```
Comment (lneto) [inline binary.c]: I would remove this warning by adding C99 into the makefile rather than modifying the original code

File: binary.c [hunk 22]
```diff
@@ -28,14 +28,16 @@
...
+#define MAX	max
+#define UCHAR_MAX	(255)
 #endif
 
-#include <sys/param.h>
-
 #include "binary.h"
 
-#define BYTE_MAX	UCHAR_MAX
+#define BYTE_MAX	(UCHAR_MAX)
```
Comment (lneto) [inline binary.c]: no need for this..

File: binary.c [hunk 23]
```diff
@@ -28,14 +28,16 @@
 #ifndef _KERNEL
 #include <limits.h>
 #else
-#include <machine/limits.h>
+#include <linux/kernel.h>
+#include "sys/endian.h"
+#define UINT64_MAX      (U64_MAX)
+#define MAX	max
```
Comment (lneto) [inline binary.c]: use the makefile for these

File: binary.c [hunk 24]
```diff
@@ -28,14 +28,16 @@
...
+#define UCHAR_MAX	(255)
```
Comment (lneto) [inline binary.c]: if not defined anywhere else.. use the makefile

File: Makefile [hunk 25]
```diff
@@ -1,16 +1,3 @@
...
-LUA_SRCS.data+=	binary.c
-
-DATA=		data.so
-LDLIBS= 	-llua  ${DATA}
-test: 		test.c ${DATA}
-
-CLEANFILES+= 	test
-
-.include <bsd.lua.mk>
+obj-y += binary.o data.o handle.o layout.o luadata.o luautil.o
```
Comment (lneto) [inline Makefile]: instead of messing with the original makefile, create a Kbuild file

File: sys/endian.h (hunk 2 above)
Comment (tammela) [inline sys/endian.h]: It's hard coded in a libc header, using grep i didnt find anything promising

**Pull request closed without merging**